frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -349,34 +341,38 @@ func toCommand(ic instructions.Command, dispatchStatesByName map[string]*dispatc
 					}
 				}
 			} else {
-				if index < 0 || index >= len(allDispatchStates) {
-					return command{}, errors.Errorf("invalid stage index %d", index)
+				stn, err = allDispatchStates.findStateByIndex(index)
+				if err != nil {
+					return command{}, err
 				}
-				stn = allDispatchStates[index]
+				ds := *newDispatchStates()
+				ds.addState(stn)
+				cmd.sources = ds
 			}
-			cmd.sources = []*dispatchState{stn}
+			ds := *newDispatchStates()
```

Comment:
Not sure about `newDispatchStates()` in here. It may be more confusing as the command sources and collection of dispatchStates for querying are not quite the same thing. You never ask for a source by index or by name. It is just a collection for iterating over - therefore I think clearer as an array.

Reply:
I did not notice command sources is just a collection for iterating over. I will update.